src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps_gen2.c

```diff
@@ -0,0 +1,190 @@
+/**
+ * @file adps_gen2.c
+ * @brief Implements the Azure DPS Gen2 communication utility.
+ *
+ * @copyright Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ */
+#include "aduc/adps_gen2.h"
+#include "aduc/config_utils.h"
+#include "aduc/logging.h"
+#include "aduc/string_c_utils.h" // ADUC_StringFormat
+#include "stdlib.h" // memset
+
+#define DEFAULT_DPS_API_VERSION "2023-02-01-preview"
+#define DEFAULT_DPS_GLOBAL_ENDPOINT "global.azure-devices-provisioning.net"
+// Default MQTT protocol version for Azure DPS Gen2 is v3.1.1 (4)
+#define DEFAULT_DPS_MQTT_PROTOCOL_VERSION 4
+
+/**
+ * @brief Read the Azure DPS Gen2 connection data from the config file.
+ *        This follow MQTT protocol describe at https://learn.microsoft.com/azure/iot/iot-mqtt-connect-to-iot-dps
+ * @param settings The Azure DPS Gen2 MQTT settings to populate. The caller is responsible for freeing the returned settings by calling FreeAzureDPS2MqttSettings function.
+ * @return true if the settings were read successfully, false otherwise.
+ */
+bool ReadAzureDPS2MqttSettings(AZURE_DPS_2_MQTT_SETTINGS* settings)
+{
+    bool result = false;
+    const ADUC_ConfigInfo* config = NULL;
+    const ADUC_AgentInfo* agent_info = NULL;
+
+    if (settings == NULL)
+    {
+        goto done;
+    }
+
+    memset(settings, 0, sizeof(*settings));
+
+    config = ADUC_ConfigInfo_GetInstance();
+    if (config == NULL)
+    {
+        goto done;
+    }
+
+    agent_info = ADUC_ConfigInfo_GetAgent(config, 0);
+    if (strcmp(agent_info->connectionType, ADUC_CONNECTION_TYPE_ADPS2_MQTT) != 0)
+    {
+        goto done;
+    }
+
+    // Read the x.506 certificate authentication data.
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "dps.caFile", &settings->mqttSettings.caFile))
+    {
+        settings->mqttSettings.caFile = NULL;
+    }
+
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "dps.certFile", &settings->mqttSettings.certFile))
+    {
+        settings->mqttSettings.certFile = NULL;
+    }
+
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "dps.keyFile", &settings->mqttSettings.keyFile))
+    {
+        settings->mqttSettings.keyFile = NULL;
+    }
+
+    // NOTE: If you use X.509 certificate authentication, the registration ID is provided by
+    // the subject common name (CN) of the device leaf (end-entity) certificate.
+    // {registration_id} in the Username field must match the common name.
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "dps.registrationId", &settings->registrationId))
+    {
+        // TODO (nox-msft) : read common name from certificate.
+        Log_Error("Cannot read registrationId field.");
+        goto done;
+    }
+
+    // For DPS connection, the clientId and registrationId fields are the same.
+    settings->mqttSettings.clientId = strdup(settings->registrationId);
```

Comment:
consider using STRING_HANDLE instead, or better yet find way to not allocate on heap each time and just copy to somewhere . same for line 112